Feature/halo exchange

middlecoff opened the pull request:
## Proposed changes

Merged feature/halo-exchange with Lynd's interpolation code. The halo-exchange code does an MPI based halo exchange between an arbitrary number of MPI tasks and provides GPTL based timings for the exchange and the csw calculations.

## Types of changes

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [ ] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-c_sw/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

middlecoff committed:
Modifications to work with interpolate.

With the hack:
    if(nTasks == 1) then
      rows = 1
      cols = 1
    endif
Passed the tests: build.sh and build.sh intel debug mpi on

I removed the hack before committing and now cmake must be modified to insure rows*cols=nTasks.

Commit:
Add namelists, update  CMake to test 1 & 4 MPI ranks

Commit:
Add namelists for 1x4 decomposition, and add respective tests

Commit:
Insert #ifdef surrounding gptl_summary call, correct c_sw_48x24_0_1x4.nl

robgpita reviewed (dismissed):
Looks good @middlecoff. I do not have much to comment on the MPI implementation as I am unfamiliar with MPI. However, I believe it is worth noting that when I added the MPI tests for 4 ranks (1x4 and 2x2 decomposition), I only tested the non-interpolated datasets and left the decomposition values at 2x2 for the interpolated datasets.

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -73,15 +93,18 @@ program sw_driver
+  ! Initialize GPTL if enabled or if this is an MPI run
 #ifdef ENABLE_GPTL
-  if (do_profile == 1) then
+  if (do_profile == 1 .or. mpi_run ) then ! GPTL automatically profiles MPI calls
```
@middlecoff - I'm assuming this is here because if the code is compiled with GPTL enabled, but `do_profile` is off, the auto profiling of the MPI calls will cause things to fail if we haven't initialized GPTL.

I wonder if the more straightforward situation is to always call the `initialize` and `finalize` routines for GPTL if `ENABLE_GPTL` is on whether `do_profile` is on or not.  The `do_profile` would then be used to toggle the actual timers on/off.  Then we can get rid of the `mpi_run` variable.

middlecoff replied:
@christopherwharrop-noaa Yes, that does sound more straightforward. I'll implement it.

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -97,6 +120,22 @@ program sw_driver
+  call get_neighbors
+
+  ! Allocate statuses for MPI_WAITALL: 0:15 because there are 8 sends and 8 receives
+  allocate( statuses(MPI_STATUS_SIZE, 0:15) )
```
Is there a better place for this?  Having it here seems out of place.

middlecoff replied:
@christopherwharrop-noaa I could put the allocate in with the other ollocates but I would need to enclose it in and IFDEF on MPI.

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -116,10 +155,35 @@ program sw_driver
+#endif
+
+  !Test the exchange
+  call test_exchange(matches)
```
It is important to do this test, but we should have the tests done separately from the kernel.

middlecoff replied:
@christopherwharrop-noaa  I agree we should do this test separately, but I think we should also do it here since takes little time and it's a standard check to make sure everything is as it should be.

christopherwharrop-noaa replied:
The issue isn't how expensive the test is, but how the kernel is organized and what it does.  I feel that the kernel code should only contain the implementation of the kernel computation and should not contain tests that check whether the computations work correctly.  You could/would not do that test in a real code that runs in production.  I feel that the kernel should just run the kernel and that it is the job of our test suite test to see if the code that does the halo exchange works correctly.  I think there needs to be a clear separation between the kernel computation and the kernel tests, particularly because this is the reference implementation that others will follow.  If we blend the the two together it makes it very confusing for others which things are the kernel and which things are the tests.

middlecoff replied:
Real codes have all kinds of error checks. In FIM and NIM MPI is encapsulated into SMS and SMS has sanity checks all over the place so it seems to me this error check is in the same sanity-check spirit.

christopherwharrop-noaa replied:
Agree that real codes have many kinds of error checks.  Disagree that this error check is in the spirit of sanity checks, particularly since (1) this check could not be done in a real code without extra communication cost, and (2) given a correct exchange algorithm, the only way this could fail is due to the network stack dropping bits.  This check is not a sanity check; it is a test to check whether the algorithm works.  Which is precisely what we need for our test suite, and which should also not be in the main body of the code.  I can not approve this PR until this test is removed from the driver and moved to the test suite.

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -73,15 +93,18 @@ program sw_driver
   read(nl_unit, nml=debug)
 
+  ! Read the MPI decomposition from the namelist
+  read(nl_unit, nml=decomposition)
```
Should this be inside of a `#ifdef ENABLE_MPI` ?

middlecoff replied:
@christopherwharrop-noaa It could be but as it is it's in the spirit of always calling GPTL init for simplicity.

christopherwharrop-noaa replied:
I'm not following you.  Why would we need to read the `decomposition` namelist if we're not using MPI?  What if the namelist doesn't have that section because someone is not using MPI?

christopherwharrop-noaa replied:
I'm not following you. Why would we need to read the decomposition namelist if we're not using MPI? What if the namelist doesn't have that section because someone is not using MPI?

middlecoff replied:
I didn't mean we needed to read the decomposition namelist I just meant it was simple and in the spirit of always calling GPTL thereby eliminating  a #IFDEF. I thought the namelistst would always have the decomposition section but I can insert the #IFDEF if you want.

christopherwharrop-noaa replied:
I'm still not following you because the decomposition has nothing to do with GPTL or whether it is turned on or not.  It is only relevant when MPI is enabled.  Since it is only used when MPI is enabled, it should be in a `#ifdef MPI_ENABLED`.

christopherwharrop-noaa reviewed (commented):
A few comments/questions.

middlecoff committed:
Implemented two of Chris' suggestions:
  1. If ENABLE_GPTL is set then always call GPTL init and finalize.
  2. Move allocation of STATUSES to the allocate routine.

middlecoff committed:
Forgot to deallocate STATUSES

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -166,14 +225,22 @@ program sw_driver
-  if (do_profile == 1) then
-    ret = gptlpr(0)
-    ret = gptlfinalize()
+  if (do_profile == 1 .or. mpi_run) then
```
Why do we need `mpi_run` here?  I thought we were getting rid of the need for that variable.

middlecoff replied:
Your suggestion simplified things but does not necessarily get rid of mpi_run. mpi_run is used here to call gptlpr to print the timings if either do_profile=1 or it's an MPI run. The case where we should certainly not call gptlpr is when ENABLE_GPTL is set and it's not an MPI run and we input do_profile /=1. 

For an MPI run, if ENABLE_GPTL is set then GPTL automatically profiles MPI but we could take the position that if do_profile/=1 then we do not want to see the automatic profile so we don't call gptlpr - is that what you had in mind?

christopherwharrop-noaa replied:
Yes, that is what I had in mind.  We should only see GPTL output if `do_profile=1`.

christopherwharrop-noaa commented on ref/src/sw_driver.F90:
```diff
@@ -21,11 +21,13 @@ program sw_driver
-  integer :: rank = 0                           ! The rank of MPI tasks
-  integer :: ierr                               ! Error return
   integer :: log_file_unit                      ! unit number for writing out log files
+  logical :: mpi_run = .false.                  ! True means this is an MPI run
```
I thought we were getting rid of this variable?

christopherwharrop-noaa reviewed (changes requested):
Please remove `mpi_run` and move the halo exchange test from the driver to a separate test program in the test suite.  Also please enclose reading of `decomposition` namelist in a `#ifdef MPI_ENABLED`

middlecoff committed:
Removed mpi_run.
Removed halo exchange test from sw_driver.F90.
Put "#ifdef MPI_ENABLED" around the read of the decompositon namelist.
One thing left do do: Create a stand alone test using the halo exchange test.

christopherwharrop-noaa committed:
Merge branch 'develop' into feature/halo-exchange

middlecoff committed:
Removed unused variable MATCHES from sw_driver.F90
Added test_exchange.F90 for error checking the exchange
Next step is to add test_exchange.F90 to cmake

middlecoff committed:
Merge branch 'feature/halo-exchange' of https://github.com/NOAA-GSL/SENA-c_sw into feature/halo-exchange

middlecoff committed:
Enhancements for style and completeness.

christopherwharrop-noaa reviewed (commented):
These final updates address all the outstanding items.  All tests are passing.  The only thing left is to properly place your new test program into the `test/` directory and hook it into the `ctest` test suite.  I will work on that.

middlecoff committed:
Forgot to update rows.

christopherwharrop-noaa committed:
Move halo exchange test to test/ directory and add to test suite

christopherwharrop-noaa committed:
Merge branch 'feature/halo-exchange' of https://github.com/NOAA-GSL/SENA-c_sw into feature/halo-exchange

christopherwharrop-noaa committed:
Fix exchange test by ensuring it uses four MPI tasks

christopherwharrop-noaa reviewed (approved):
This is ready for merging.

robgpita reviewed (commented):
@christopherwharrop-noaa @middlecoff Great work!
